Sources/StreamChat/StateLayer/MessageSearchState+Observer.swift

```diff
@@ -0,0 +1,62 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+@available(iOS 13.0, *)
+extension MessageSearchState {
+    final class Observer {
+        private let database: DatabaseContainer
+        private var messagesObserver: BackgroundListDatabaseObserver<ChatMessage, MessageDTO>?
+
+        init(database: DatabaseContainer) {
+            self.database = database
+        }
+        
+        deinit {
+            guard let query else { return }
+            database.write { $0.deleteQuery(query) }
+        }
+        
+        struct Handlers {
+            let messagesDidChange: (StreamCollection<ChatMessage>) async -> Void
+        }
+        
+        private var handlers: Handlers?
+        
+        func start(with handlers: Handlers) {
+            self.handlers = handlers
+        }
+        
+        var query: MessageSearchQuery? {
+            didSet {
+                reset(to: query)
+            }
+        }
+        
+        private func reset(to query: MessageSearchQuery?) {
+            if let query {
+                messagesObserver = BackgroundListDatabaseObserver(
+                    context: database.backgroundReadOnlyContext,
+                    fetchRequest: MessageDTO.messagesFetchRequest(for: query),
+                    itemCreator: { try $0.asModel() as ChatMessage },
+                    sorting: []
+                )
```

Comment:
In general I am not happy about this part but we need to do the work of supporting MessagePayload>ChatMessage conversion first. Slightly silly since results are API driven and we do not really use the CD storage for search. But that is the current situation.